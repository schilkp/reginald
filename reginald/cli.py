from collections.abc import Callable
from dataclasses import dataclass
from typing import List

from typing_extensions import Self

from reginald.datamodel import RegisterMap


@dataclass
class CLI:
    input_file: str
    output_path: str
    generator: Callable[[RegisterMap, Self], None]
    generator_args: List[str]
