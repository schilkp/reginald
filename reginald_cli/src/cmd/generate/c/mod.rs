pub mod funcpack;
pub mod macromap;
