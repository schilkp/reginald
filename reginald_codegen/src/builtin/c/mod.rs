use std::fmt::Write;

use crate::{
    error::Error,
    regmap::{Docs, Layout, TypeBitwidth},
    utils::str_pad_to_length,
};
use lazy_static::lazy_static;
use regex::Regex;

pub mod funcpack;
pub mod macromap;

fn c_macro(s: &str) -> String {
    c_sanitize(&s.to_uppercase())
}

fn c_code(s: &str) -> String {
    c_sanitize(&s.to_lowercase())
}

lazy_static! {
    static ref C_SANITIZE_REGEX: Regex = Regex::new(r"[^_a-zA-Z0-9]").unwrap();
}

fn c_sanitize(s: &str) -> String {
    C_SANITIZE_REGEX.replace_all(s, "_").into()
}

fn c_fitting_unsigned_type(width: TypeBitwidth) -> Result<String, Error> {
    match width {
        1..=8 => Ok("uint8_t".to_string()),
        9..=16 => Ok("uint16_t".to_string()),
        17..=32 => Ok("uint32_t".to_string()),
        33..=64 => Ok("uint64_t".to_string()),
        _ => Err(Error::GeneratorError(format!("Cannot represent {width}-bit wide value as C type!"))),
    }
}

fn c_section_header_comment(title: &str) -> String {
    let lbar = str_pad_to_length("// ", '=', 80);
    let lcontent = str_pad_to_length(&format!("// ==== {title} "), '=', 80);

    format!("{lbar}\n{lcontent}\n{lbar}")
}

fn c_generate_section_header_comment(out: &mut dyn Write, title: &str) -> Result<(), Error> {
    writeln!(out, "{}", c_section_header_comment(title))?;
    Ok(())
}

fn c_header_comment(title: &str) -> String {
    str_pad_to_length(&format!("// ---- {title} "), '-', 80)
}

fn c_generate_header_comment(out: &mut dyn Write, title: &str) -> Result<(), Error> {
    writeln!(out, "{}", c_header_comment(title))?;
    Ok(())
}

fn c_generate_doxy_comment(
    out: &mut dyn Write,
    docs: &Docs,
    prefix: &str,
    mut nodes: Vec<(String, String)>,
) -> Result<(), Error> {
    if let Some(brief) = &docs.brief {
        nodes.push((String::from("brief"), brief.to_owned()));
    }

    nodes.sort_by_key(|x| match x.0.as_str() {
        "name" => 0,
        "brief" => 1,
        "warning" => 2,
        "note" => 3,
        "returns" => 4,
        _ => 5,
    });

    if nodes.is_empty() && docs.doc.is_none() {
        Ok(())
    } else if nodes.len() == 1 && docs.doc.is_none() {
        writeln!(out, "{prefix}/** @{} {} */", nodes[0].0, nodes[0].1)?;
        Ok(())
    } else {
        writeln!(out, "{prefix}/**")?;
        for (node_name, node) in &nodes {
            writeln!(out, "{prefix} * @{node_name} {node}")?
        }
        if let Some(doc) = &docs.doc {
            for line in doc.lines() {
                writeln!(out, "{prefix} * {line}")?;
            }
        }
        writeln!(out, "{prefix} */")?;
        Ok(())
    }
}

fn c_layout_overview_comment(layout: &Layout) -> String {
    layout
        .overview_text(false)
        .lines()
        .map(|x| String::from("//  ") + x)
        .collect::<Vec<String>>()
        .join("\n")
}
