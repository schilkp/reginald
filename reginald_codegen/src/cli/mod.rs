mod cmd;
mod diff;

use std::process::ExitCode;

use clap::Parser;

#[derive(Parser, Debug)]
#[command(version, about)]
#[command(name = "reginald")]
enum Cli {
    Gen(cmd::gen::Command),
    Completion(cmd::completion::Command),
    Tool(cmd::tool::Command),
}

pub fn cli_main() -> ExitCode {
    let cli = Cli::parse();

    let err = match cli {
        Cli::Gen(gen) => cmd::gen::cmd(gen),
        Cli::Completion(c) => cmd::completion::cmd(c),
        Cli::Tool(tool) => cmd::tool::cmd(tool),
    };

    match err {
        Ok(_) => ExitCode::SUCCESS,
        Err(err) => {
            eprintln!("{}", err);
            ExitCode::FAILURE
        }
    }
}
