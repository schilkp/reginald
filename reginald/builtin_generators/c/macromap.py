import os
from typing import List, Tuple, Union

from reginald.cli import CLI
from reginald.datamodel import *
from reginald.generator import OutputGenerator
from reginald.utils import c_sanitize, str_pad_to_length


class Generator(OutputGenerator):
    @classmethod
    def description(cls):
        return "TODO"

    @classmethod
    def generate(cls, rmap: RegisterMap, cli: CLI):

        mapname = rmap.map_name
        mapname_macro = c_sanitize(mapname).upper()

        out = []

        out.append(f"/*")
        out.append(f"* {mapname} Register Map.")
        out.append(f"* Note: Do not edit: Generated using Reginald.")
        out.append(f"*/")
        out.append(f"")
        out.append(f"#ifndef {mapname_macro}_REG_H_")
        out.append(f"#define {mapname_macro}_REG_H_")
        out.append(f"")

        # combine physical registers and register templates:
        registers = {}  # type: Dict[str, Union[Register, RegisterTemplate]]

        for register in rmap.registers.values():
            registers[register.name] = register

        for block_template in rmap.register_block_templates.values():
            for template in block_template.registers.values():
                registers[block_template.name + template.name] = template

        for reg_name, reg in registers.items():
            reg_name = c_sanitize(reg_name)

            out.append(str_pad_to_length(f"// ==== {reg_name} ", "=", 80))

            out.extend(reg.docs.multi_line(prefix="// "))

            # Generate all defines, keeping comments seperate for now:

            defines = []  # type: List[Tuple[str, str, str]]

            register_prefix = f"{mapname_macro}__REG_{reg_name}"
            if isinstance(reg, Register):
                if reg.adr is not None:
                    docstr = f"({reg.docs.brief})" if reg.docs.brief is not None else ""
                    defines.append((f"#define {register_prefix}", f"(0x{reg.adr:02X}U)", f"// Register Address {docstr}"))
            else:
                docstr = f"({reg.docs.brief})" if reg.docs.brief is not None else ""
                defines.append((f"#define {register_prefix}", f"(0x{reg.offset:02X}U)", f"// Register Offset {docstr}"))

            if reg.reset_val is not None:
                defines.append((f"#define {register_prefix}__RESET", f"(0x{reg.reset_val:02X}U)", f"// Reset Value"))

            if reg.always_write is not None:
                mask = reg.always_write.bits.get_bitmask()
                value = reg.always_write.value
                defines.append((f"#define {register_prefix}__ALWAYSWRITE_MASK", f"(0x{mask:02X}U)", f"// 'Always write' bit mask"))
                defines.append((f"#define {register_prefix}__ALWAYSWRITE_VALUE", f"(0x{value:02X}U)", f"// 'Always write' value"))

            for field in reg.fields.values():
                docstr = f"({field.docs.brief})" if field.docs.brief is not None else ""
                field_name = c_sanitize(field.name)
                field_prefix = f"{register_prefix}__FIELD_{field_name}"
                mask = field.bits.get_bitmask()
                defines.append((f"#define {field_prefix}", f"(0x{mask:02X}U)", f"// Field Mask {docstr}"))

                if field.enum is not None:
                    for const in field.enum.entries.values():
                        docstr = f"({const.docs.brief})" if const.docs.brief is not None else ""
                        const_name = c_sanitize(const.name)
                        const_prefix = f"{field_prefix}__CONST_{const_name}"
                        defines.append((f"#define {const_prefix}", f"(0x{const.value:02X}U)", f"// Constant {docstr}"))

            # Align values and comments:
            define_max_len = max([len(d[0]) for d in defines]) + 1
            val_max_len = max([len(d[1]) for d in defines]) + 1

            for define, val, comment in defines:
                define = define + (" " * (define_max_len - len(define)))
                val = (" " * (val_max_len - len(val))) + val
                out.append(define+val+" "+comment)

            out.append(f"")

        out.append(f"")
        out.append(f"#endif /* {mapname_macro}_REG_H_ */")

        output_file = os.path.join(cli.output_path, f"{rmap.map_name.lower()}_regs.h")
        with open(output_file, 'w') as outfile:
            outfile.write("\n".join(out)+"\n")
        print(f"Generated {output_file}...")
