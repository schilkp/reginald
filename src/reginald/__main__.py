import sys

from reginald.cli import parse_args
from reginald.error import ReginaldException
from reginald.input.convert_yaml import YAMLConverter
from reginald.input.parse_yaml import YAML_RegisterMap
from reginald.input.validate_map import MapValidator


def main():

    try:

        # Parse command line args:
        cli, generator = parse_args()

        # Open, parse, and validate input file:
        r = YAML_RegisterMap.from_yaml_file(cli.input_file)
        r = YAMLConverter(r).convert()
        MapValidator(r).validate()

        # Generate output using selected generator:
        generator.generate(r, cli.input_file, cli.output_file, cli.generator_args)

    except ReginaldException as e:
        print(e, file=sys.stderr)
        exit(-1)

    exit(0)


if __name__ == '__main__':
    main()
