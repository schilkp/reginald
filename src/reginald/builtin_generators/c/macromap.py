from reginald.jinja2_generator import BuiltinJinjaGenerator


class Generator(BuiltinJinjaGenerator):
    def __init__(self):
        super().__init__(desc="C header with traditional register and field macros.", template_name="c/macromap.h")
