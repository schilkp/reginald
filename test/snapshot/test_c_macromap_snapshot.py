import os

import reginald.builtin_generators.c.macromap as builtin_c_macromap
from reginald.__main__ import generate
from reginald.cli import CLI

SNAPSHOT_DIR = os.path.dirname(__file__)


def test_c_macromap_8b(snapshot):
    generator = builtin_c_macromap.Generator()
    cli = CLI(input_file=os.path.join(SNAPSHOT_DIR, "input_8b.yaml"),
              output_file="c_macromap_snapshot_8b.h",
              generator_args=[],
              verify=False)
    output = "\n".join(generate(cli, generator))
    assert output == snapshot


def test_c_macromap_9b(snapshot):
    generator = builtin_c_macromap.Generator()
    cli = CLI(input_file=os.path.join(SNAPSHOT_DIR, "input_9b.yaml"),
              output_file="c_macromap_snapshot_9b.h",
              generator_args=[],
              verify=False)
    output = "\n".join(generate(cli, generator))
    assert output == snapshot


def test_c_macromap_32b(snapshot):
    generator = builtin_c_macromap.Generator()
    cli = CLI(input_file=os.path.join(SNAPSHOT_DIR, "input_32b.yaml"),
              output_file="c_macromap_snapshot_32b.h",
              generator_args=[],
              verify=False)
    output = "\n".join(generate(cli, generator))
    assert output == snapshot


def test_c_macromap_64b(snapshot):
    generator = builtin_c_macromap.Generator()
    cli = CLI(input_file=os.path.join(SNAPSHOT_DIR, "input_64b.yaml"),
              output_file="c_macromap_snapshot_64b.h",
              generator_args=[],
              verify=False)
    output = "\n".join(generate(cli, generator))
    assert output == snapshot
