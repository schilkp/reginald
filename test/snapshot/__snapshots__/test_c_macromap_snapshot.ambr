# serializer version: 1
# name: test_c_macromap_32b
  '''
  /*
   * Snapshotter32B Register Map.
   * Note: Do not edit: Generated using Reginald.
   */
  #ifndef C_MACROMAP_SNAPSHOT_32B_H_
  #define C_MACROMAP_SNAPSHOT_32B_H_
  
  
  // === BASIC ===================================================================
  
  #define SNAPSHOTTER32B__REG_BASIC (0x0U) // Register address.
  #define SNAPSHOTTER32B__REG_BASIC__FIELD_FIELD0 (0xfffffe00U) // Field mask.
  #define SNAPSHOTTER32B__REG_BASIC__FIELD_FIELD1 (0xc0U) // Field mask.
  #define SNAPSHOTTER32B__REG_BASIC__FIELD_FIELD2 (0x20U) // Field mask.
  #define SNAPSHOTTER32B__REG_BASIC__FIELD_FIELD3 (0x18U) // Field mask.
  #define SNAPSHOTTER32B__REG_BASIC__FIELD_FIELD4 (0x6U) // Field mask.
  #define SNAPSHOTTER32B__REG_BASIC__FIELD_FIELD5 (0x1U) // Field mask.
  
  // === COMPLICATED =============================================================
  // Status.
  // More extensive documentation.
  
  #define SNAPSHOTTER32B__REG_COMPLICATED (0xfU) // Register address "Status.".
  #define SNAPSHOTTER32B__REG_COMPLICATED__RESET (0xffffffffU) // Reset value.
  #define SNAPSHOTTER32B__REG_COMPLICATED__ALWAYS_WRITE_MASK (0xc30U) // Always-write bit mask.
  #define SNAPSHOTTER32B__REG_COMPLICATED__ALWAYS_WRITE_VAL (0xc10U) //  Always-write value.
  #define SNAPSHOTTER32B__REG_COMPLICATED__FIELD_FIELD1 (0xc0U) // Field mask.
  #define SNAPSHOTTER32B__REG_COMPLICATED__FIELD_FIELD1__CONST_COOL (0x1U) // Constant "All cool.".
  #define SNAPSHOTTER32B__REG_COMPLICATED__FIELD_FIELD1__CONST_NOT_COOL (0x2U) // Constant "All not so cool.".
  #define SNAPSHOTTER32B__REG_COMPLICATED__FIELD_FIELD1__CONST_HOT (0x3U) // Constant "All pretty hot.".
  #define SNAPSHOTTER32B__REG_COMPLICATED__FIELD_FIELD2 (0xfU) // Field mask "Another cool field.".
  #define SNAPSHOTTER32B__REG_COMPLICATED__FIELD_FIELD2__CONST_EN (0x3U) // Constant "Is on.".
  
  // === CHn_CNFG ================================================================
  
  #define SNAPSHOTTER32B__REG_CH1_CNFG (0x10U) // Register address.
  #define SNAPSHOTTER32B__REG_CH2_CNFG (0x20U) // Register address.
  #define SNAPSHOTTER32B__REG_CH3_CNFG (0x30U) // Register address.
  #define SNAPSHOTTER32B__REG_CHN_CNFG__OFFSET (0x0U) // Offset of CHn_CNFG register from start of CHn block.
  #define SNAPSHOTTER32B__REG_CHN_CNFG__RESET (0x0U) // Reset value.
  #define SNAPSHOTTER32B__REG_CHN_CNFG__FIELD_Y (0xc0000000U) // Field mask.
  #define SNAPSHOTTER32B__REG_CHN_CNFG__FIELD_A (0x1U) // Field mask "Controls value A".
  #define SNAPSHOTTER32B__REG_CHN_CNFG__FIELD_A__CONST_ON (0x1U) // Constant.
  #define SNAPSHOTTER32B__REG_CHN_CNFG__FIELD_A__CONST_OFF (0x0U) // Constant.
  
  // === CHn_STATUS ==============================================================
  
  #define SNAPSHOTTER32B__REG_CH1_STATUS (0x11U) // Register address.
  #define SNAPSHOTTER32B__REG_CH2_STATUS (0x21U) // Register address.
  #define SNAPSHOTTER32B__REG_CH3_STATUS (0x31U) // Register address.
  #define SNAPSHOTTER32B__REG_CHN_STATUS__OFFSET (0x1U) // Offset of CHn_STATUS register from start of CHn block.
  #define SNAPSHOTTER32B__REG_CHN_STATUS__RESET (0x0U) // Reset value.
  #define SNAPSHOTTER32B__REG_CHN_STATUS__FIELD_STATUS (0x7U) // Field mask "Status".
  #define SNAPSHOTTER32B__REG_CHN_STATUS__FIELD_STATUS__CONST_COOL (0x1U) // Constant "All cool.".
  #define SNAPSHOTTER32B__REG_CHN_STATUS__FIELD_STATUS__CONST_NOT_COOL (0x2U) // Constant "All not so cool.".
  #define SNAPSHOTTER32B__REG_CHN_STATUS__FIELD_STATUS__CONST_HOT (0x3U) // Constant "All pretty hot.".
  
  // === CHn =====================================================================
  
  #define SNAPSHOTTER32B__BLOCK_CHN__START_CH1 (0x10U) // Start of CH1 register block.
  #define SNAPSHOTTER32B__BLOCK_CHN__START_CH2 (0x20U) // Start of CH2 register block.
  #define SNAPSHOTTER32B__BLOCK_CHN__START_CH3 (0x30U) // Start of CH3 register block.
  
  #endif /* C_MACROMAP_SNAPSHOT_32B_H_ */
  '''
# ---
# name: test_c_macromap_64b
  '''
  /*
   * Snapshotter64B Register Map.
   * Note: Do not edit: Generated using Reginald.
   */
  #ifndef C_MACROMAP_SNAPSHOT_64B_H_
  #define C_MACROMAP_SNAPSHOT_64B_H_
  
  
  // === BASIC ===================================================================
  
  #define SNAPSHOTTER64B__REG_BASIC (0x0U) // Register address.
  #define SNAPSHOTTER64B__REG_BASIC__FIELD_FIELD0 (0xfffffffffffffe00U) // Field mask.
  #define SNAPSHOTTER64B__REG_BASIC__FIELD_FIELD1 (0xc0U) // Field mask.
  #define SNAPSHOTTER64B__REG_BASIC__FIELD_FIELD2 (0x20U) // Field mask.
  #define SNAPSHOTTER64B__REG_BASIC__FIELD_FIELD3 (0x18U) // Field mask.
  #define SNAPSHOTTER64B__REG_BASIC__FIELD_FIELD4 (0x6U) // Field mask.
  #define SNAPSHOTTER64B__REG_BASIC__FIELD_FIELD5 (0x1U) // Field mask.
  
  // === COMPLICATED =============================================================
  // Status.
  // More extensive documentation.
  
  #define SNAPSHOTTER64B__REG_COMPLICATED (0xfU) // Register address "Status.".
  #define SNAPSHOTTER64B__REG_COMPLICATED__RESET (0xffffffffU) // Reset value.
  #define SNAPSHOTTER64B__REG_COMPLICATED__ALWAYS_WRITE_MASK (0xc30U) // Always-write bit mask.
  #define SNAPSHOTTER64B__REG_COMPLICATED__ALWAYS_WRITE_VAL (0xc10U) //  Always-write value.
  #define SNAPSHOTTER64B__REG_COMPLICATED__FIELD_FIELD1 (0xc0U) // Field mask.
  #define SNAPSHOTTER64B__REG_COMPLICATED__FIELD_FIELD1__CONST_COOL (0x1U) // Constant "All cool.".
  #define SNAPSHOTTER64B__REG_COMPLICATED__FIELD_FIELD1__CONST_NOT_COOL (0x2U) // Constant "All not so cool.".
  #define SNAPSHOTTER64B__REG_COMPLICATED__FIELD_FIELD1__CONST_HOT (0x3U) // Constant "All pretty hot.".
  #define SNAPSHOTTER64B__REG_COMPLICATED__FIELD_FIELD2 (0xfU) // Field mask "Another cool field.".
  #define SNAPSHOTTER64B__REG_COMPLICATED__FIELD_FIELD2__CONST_EN (0x3U) // Constant "Is on.".
  
  // === CHn_CNFG ================================================================
  
  #define SNAPSHOTTER64B__REG_CH1_CNFG (0x10U) // Register address.
  #define SNAPSHOTTER64B__REG_CH2_CNFG (0x20U) // Register address.
  #define SNAPSHOTTER64B__REG_CH3_CNFG (0x30U) // Register address.
  #define SNAPSHOTTER64B__REG_CHN_CNFG__OFFSET (0x0U) // Offset of CHn_CNFG register from start of CHn block.
  #define SNAPSHOTTER64B__REG_CHN_CNFG__RESET (0x0U) // Reset value.
  #define SNAPSHOTTER64B__REG_CHN_CNFG__FIELD_Y (0xffffffffc0000000U) // Field mask.
  #define SNAPSHOTTER64B__REG_CHN_CNFG__FIELD_A (0x1U) // Field mask "Controls value A".
  #define SNAPSHOTTER64B__REG_CHN_CNFG__FIELD_A__CONST_ON (0x1U) // Constant.
  #define SNAPSHOTTER64B__REG_CHN_CNFG__FIELD_A__CONST_OFF (0x0U) // Constant.
  
  // === CHn_STATUS ==============================================================
  
  #define SNAPSHOTTER64B__REG_CH1_STATUS (0x11U) // Register address.
  #define SNAPSHOTTER64B__REG_CH2_STATUS (0x21U) // Register address.
  #define SNAPSHOTTER64B__REG_CH3_STATUS (0x31U) // Register address.
  #define SNAPSHOTTER64B__REG_CHN_STATUS__OFFSET (0x1U) // Offset of CHn_STATUS register from start of CHn block.
  #define SNAPSHOTTER64B__REG_CHN_STATUS__RESET (0x0U) // Reset value.
  #define SNAPSHOTTER64B__REG_CHN_STATUS__FIELD_STATUS (0x7U) // Field mask "Status".
  #define SNAPSHOTTER64B__REG_CHN_STATUS__FIELD_STATUS__CONST_COOL (0x1U) // Constant "All cool.".
  #define SNAPSHOTTER64B__REG_CHN_STATUS__FIELD_STATUS__CONST_NOT_COOL (0x2U) // Constant "All not so cool.".
  #define SNAPSHOTTER64B__REG_CHN_STATUS__FIELD_STATUS__CONST_HOT (0x3U) // Constant "All pretty hot.".
  
  // === CHn =====================================================================
  
  #define SNAPSHOTTER64B__BLOCK_CHN__START_CH1 (0x10U) // Start of CH1 register block.
  #define SNAPSHOTTER64B__BLOCK_CHN__START_CH2 (0x20U) // Start of CH2 register block.
  #define SNAPSHOTTER64B__BLOCK_CHN__START_CH3 (0x30U) // Start of CH3 register block.
  
  #endif /* C_MACROMAP_SNAPSHOT_64B_H_ */
  '''
# ---
# name: test_c_macromap_8b
  '''
  /*
   * Snapshotter8B Register Map.
   * Note: Do not edit: Generated using Reginald.
   */
  #ifndef C_MACROMAP_SNAPSHOT_8B_H_
  #define C_MACROMAP_SNAPSHOT_8B_H_
  
  
  // === BASIC ===================================================================
  
  #define SNAPSHOTTER8B__REG_BASIC (0x0U) // Register address.
  #define SNAPSHOTTER8B__REG_BASIC__FIELD_FIELD1 (0xc0U) // Field mask.
  #define SNAPSHOTTER8B__REG_BASIC__FIELD_FIELD2 (0x20U) // Field mask.
  #define SNAPSHOTTER8B__REG_BASIC__FIELD_FIELD3 (0x18U) // Field mask.
  #define SNAPSHOTTER8B__REG_BASIC__FIELD_FIELD4 (0x6U) // Field mask.
  #define SNAPSHOTTER8B__REG_BASIC__FIELD_FIELD5 (0x1U) // Field mask.
  
  // === COMPLICATED =============================================================
  // Status.
  // More extensive documentation.
  
  #define SNAPSHOTTER8B__REG_COMPLICATED (0xfU) // Register address "Status.".
  #define SNAPSHOTTER8B__REG_COMPLICATED__RESET (0x0U) // Reset value.
  #define SNAPSHOTTER8B__REG_COMPLICATED__ALWAYS_WRITE_MASK (0x30U) // Always-write bit mask.
  #define SNAPSHOTTER8B__REG_COMPLICATED__ALWAYS_WRITE_VAL (0x10U) //  Always-write value.
  #define SNAPSHOTTER8B__REG_COMPLICATED__FIELD_FIELD1 (0xc0U) // Field mask.
  #define SNAPSHOTTER8B__REG_COMPLICATED__FIELD_FIELD1__CONST_COOL (0x1U) // Constant "All cool.".
  #define SNAPSHOTTER8B__REG_COMPLICATED__FIELD_FIELD1__CONST_NOT_COOL (0x2U) // Constant "All not so cool.".
  #define SNAPSHOTTER8B__REG_COMPLICATED__FIELD_FIELD1__CONST_HOT (0x3U) // Constant "All pretty hot.".
  #define SNAPSHOTTER8B__REG_COMPLICATED__FIELD_FIELD2 (0xfU) // Field mask "Another cool field.".
  #define SNAPSHOTTER8B__REG_COMPLICATED__FIELD_FIELD2__CONST_EN (0x3U) // Constant "Is on.".
  
  // === CHn_CNFG ================================================================
  
  #define SNAPSHOTTER8B__REG_CH1_CNFG (0x10U) // Register address.
  #define SNAPSHOTTER8B__REG_CH2_CNFG (0x20U) // Register address.
  #define SNAPSHOTTER8B__REG_CH3_CNFG (0x30U) // Register address.
  #define SNAPSHOTTER8B__REG_CHN_CNFG__OFFSET (0x0U) // Offset of CHn_CNFG register from start of CHn block.
  #define SNAPSHOTTER8B__REG_CHN_CNFG__RESET (0x0U) // Reset value.
  #define SNAPSHOTTER8B__REG_CHN_CNFG__FIELD_A (0x1U) // Field mask "Controls value A".
  #define SNAPSHOTTER8B__REG_CHN_CNFG__FIELD_A__CONST_ON (0x1U) // Constant.
  #define SNAPSHOTTER8B__REG_CHN_CNFG__FIELD_A__CONST_OFF (0x0U) // Constant.
  
  // === CHn_STATUS ==============================================================
  
  #define SNAPSHOTTER8B__REG_CH1_STATUS (0x11U) // Register address.
  #define SNAPSHOTTER8B__REG_CH2_STATUS (0x21U) // Register address.
  #define SNAPSHOTTER8B__REG_CH3_STATUS (0x31U) // Register address.
  #define SNAPSHOTTER8B__REG_CHN_STATUS__OFFSET (0x1U) // Offset of CHn_STATUS register from start of CHn block.
  #define SNAPSHOTTER8B__REG_CHN_STATUS__RESET (0x0U) // Reset value.
  #define SNAPSHOTTER8B__REG_CHN_STATUS__FIELD_STATUS (0x7U) // Field mask "Status".
  #define SNAPSHOTTER8B__REG_CHN_STATUS__FIELD_STATUS__CONST_COOL (0x1U) // Constant "All cool.".
  #define SNAPSHOTTER8B__REG_CHN_STATUS__FIELD_STATUS__CONST_NOT_COOL (0x2U) // Constant "All not so cool.".
  #define SNAPSHOTTER8B__REG_CHN_STATUS__FIELD_STATUS__CONST_HOT (0x3U) // Constant "All pretty hot.".
  
  // === CHn =====================================================================
  
  #define SNAPSHOTTER8B__BLOCK_CHN__START_CH1 (0x10U) // Start of CH1 register block.
  #define SNAPSHOTTER8B__BLOCK_CHN__START_CH2 (0x20U) // Start of CH2 register block.
  #define SNAPSHOTTER8B__BLOCK_CHN__START_CH3 (0x30U) // Start of CH3 register block.
  
  #endif /* C_MACROMAP_SNAPSHOT_8B_H_ */
  '''
# ---
# name: test_c_macromap_9b
  '''
  /*
   * Snapshotter9B Register Map.
   * Note: Do not edit: Generated using Reginald.
   */
  #ifndef C_MACROMAP_SNAPSHOT_9B_H_
  #define C_MACROMAP_SNAPSHOT_9B_H_
  
  
  // === BASIC ===================================================================
  
  #define SNAPSHOTTER9B__REG_BASIC (0x0U) // Register address.
  #define SNAPSHOTTER9B__REG_BASIC__FIELD_FIELD0 (0x100U) // Field mask.
  #define SNAPSHOTTER9B__REG_BASIC__FIELD_FIELD1 (0xc0U) // Field mask.
  #define SNAPSHOTTER9B__REG_BASIC__FIELD_FIELD2 (0x20U) // Field mask.
  #define SNAPSHOTTER9B__REG_BASIC__FIELD_FIELD3 (0x18U) // Field mask.
  #define SNAPSHOTTER9B__REG_BASIC__FIELD_FIELD4 (0x6U) // Field mask.
  #define SNAPSHOTTER9B__REG_BASIC__FIELD_FIELD5 (0x1U) // Field mask.
  
  // === COMPLICATED =============================================================
  // Status.
  // More extensive documentation.
  
  #define SNAPSHOTTER9B__REG_COMPLICATED (0xfU) // Register address "Status.".
  #define SNAPSHOTTER9B__REG_COMPLICATED__RESET (0x0U) // Reset value.
  #define SNAPSHOTTER9B__REG_COMPLICATED__ALWAYS_WRITE_MASK (0x30U) // Always-write bit mask.
  #define SNAPSHOTTER9B__REG_COMPLICATED__ALWAYS_WRITE_VAL (0x10U) //  Always-write value.
  #define SNAPSHOTTER9B__REG_COMPLICATED__FIELD_FIELD1 (0x1c0U) // Field mask.
  #define SNAPSHOTTER9B__REG_COMPLICATED__FIELD_FIELD1__CONST_COOL (0x1U) // Constant "All cool.".
  #define SNAPSHOTTER9B__REG_COMPLICATED__FIELD_FIELD1__CONST_NOT_COOL (0x2U) // Constant "All not so cool.".
  #define SNAPSHOTTER9B__REG_COMPLICATED__FIELD_FIELD1__CONST_HOT (0x3U) // Constant "All pretty hot.".
  #define SNAPSHOTTER9B__REG_COMPLICATED__FIELD_FIELD2 (0xfU) // Field mask "Another cool field.".
  #define SNAPSHOTTER9B__REG_COMPLICATED__FIELD_FIELD2__CONST_EN (0x3U) // Constant "Is on.".
  
  // === CHn_CNFG ================================================================
  
  #define SNAPSHOTTER9B__REG_CH1_CNFG (0x10U) // Register address.
  #define SNAPSHOTTER9B__REG_CH2_CNFG (0x20U) // Register address.
  #define SNAPSHOTTER9B__REG_CH3_CNFG (0x30U) // Register address.
  #define SNAPSHOTTER9B__REG_CHN_CNFG__OFFSET (0x0U) // Offset of CHn_CNFG register from start of CHn block.
  #define SNAPSHOTTER9B__REG_CHN_CNFG__RESET (0x0U) // Reset value.
  #define SNAPSHOTTER9B__REG_CHN_CNFG__FIELD_A (0x180U) // Field mask "Controls value A".
  #define SNAPSHOTTER9B__REG_CHN_CNFG__FIELD_A__CONST_ON (0x1U) // Constant.
  #define SNAPSHOTTER9B__REG_CHN_CNFG__FIELD_A__CONST_OFF (0x0U) // Constant.
  
  // === CHn_STATUS ==============================================================
  
  #define SNAPSHOTTER9B__REG_CH1_STATUS (0x11U) // Register address.
  #define SNAPSHOTTER9B__REG_CH2_STATUS (0x21U) // Register address.
  #define SNAPSHOTTER9B__REG_CH3_STATUS (0x31U) // Register address.
  #define SNAPSHOTTER9B__REG_CHN_STATUS__OFFSET (0x1U) // Offset of CHn_STATUS register from start of CHn block.
  #define SNAPSHOTTER9B__REG_CHN_STATUS__RESET (0x0U) // Reset value.
  #define SNAPSHOTTER9B__REG_CHN_STATUS__FIELD_STATUS (0x7U) // Field mask "Status".
  #define SNAPSHOTTER9B__REG_CHN_STATUS__FIELD_STATUS__CONST_COOL (0x1U) // Constant "All cool.".
  #define SNAPSHOTTER9B__REG_CHN_STATUS__FIELD_STATUS__CONST_NOT_COOL (0x2U) // Constant "All not so cool.".
  #define SNAPSHOTTER9B__REG_CHN_STATUS__FIELD_STATUS__CONST_HOT (0x3U) // Constant "All pretty hot.".
  
  // === CHn =====================================================================
  
  #define SNAPSHOTTER9B__BLOCK_CHN__START_CH1 (0x10U) // Start of CH1 register block.
  #define SNAPSHOTTER9B__BLOCK_CHN__START_CH2 (0x20U) // Start of CH2 register block.
  #define SNAPSHOTTER9B__BLOCK_CHN__START_CH3 (0x30U) // Start of CH3 register block.
  
  #endif /* C_MACROMAP_SNAPSHOT_9B_H_ */
  '''
# ---
